Google.Api.Generator/Generation/ServiceBuilderCodeGenerator.cs

```diff
@@ -63,27 +68,54 @@ private ClassDeclarationSyntax Generate()
             AutoProperty(Public, _ctx.Type(_svc.SettingsTyp), "Settings", hasSetter: true)
                 .WithXmlDoc(XmlDoc.Summary("The settings to use for RPCs, or ", null, " for the default settings."));
 
+        private MethodDeclarationSyntax InterceptBuild() => PartialMethod("InterceptBuild")(Parameter(_ctx.Type(_svc.ClientAbstractTyp), "client").Ref());
+
+        private MethodDeclarationSyntax InterceptBuildAsync() => PartialMethod("InterceptBuildAsync")(
+            Parameter(_ctx.Type<CancellationToken>(), "cancellationToken"),
+            Parameter(_ctx.Type(Typ.Generic(typeof(Task<>), _svc.ClientAbstractTyp)), "task").Ref());
+
         private MethodDeclarationSyntax Build()
         {
-            var callInvoker = Local(_ctx.Type<CallInvoker>(), "callInvoker");
+            var client = Local(_ctx.Type(_svc.ClientAbstractTyp), "client");
             return Method(Public | Override, _ctx.Type(_svc.ClientAbstractTyp), "Build")()
                 .WithBody(
-                    This.Call("Validate")(),
-                    callInvoker.WithInitializer(This.Call("CreateCallInvoker")()),
-                    Return(_ctx.Type(_svc.ClientAbstractTyp).Call("Create")(callInvoker, Settings())))
+                    client.WithInitializer(Null),
+                    This.Call("InterceptBuild")(Ref(client)),
+                    Return(BinaryExpression(SyntaxKind.CoalesceExpression, IdentifierName("client"), This.Call("BuildImpl")())))
```

Comment:
Please use `client.NullCoalesce(This.Call("BuildImpl")())`

`NullCoalesce` is an extension method defined in `RoslynExtensions.cs`
(and then the `using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory` can go too)

Reply:
Hooray. I was surprised to have to do this, but the discoverability in Roslyn isn't great.

Reply:
Ah - it's not quite this simple because "client" is a statement, not an expression syntax. But it's not too bad...